### PR TITLE
Changelog window is alive again

### LINKED ISSUE #5665: Integrate CHANGELOG.md into the game
### Description
We implemented a system to keep track of changes in a rather chaotical git history, by manually declaring them on our PR's. I suggest a way to take that effort and make it shine in game.

### Requirements
- Take changes from CHANGELOG.md and present them in game, in a easy to read but unobstructive way.

### Implementation suggestions
- Make each build copy changes from CHANGELOG.md to a file in the project folder, somewhere inside ``StreammingAssets``.
- Make the build compare the copy with whatever already exists in the game project folder, so it can group changes by **"what's new on latest build"** (changes that are not present in the previous file) and **"older changes"** (changes that were already present).
- Present them in a new tab of the floating window that appears on Lobby
ugly mockup

### PR DESCRIPTION
**Pull request by MaxIsJoe**
Dae will be happy now

closes #5665

### Notes:
Uses the already existing old UI that I did not know it existed which thank god because I dont want to mess around with ui too much

### Changelog:

CL: [New] Changelogs appear in-game again.

### CODE REVIEW

**Comment by MaxIsJoe** [comment]
Gillies just needs to reverse the order of the json list or give me the option to reverse it via a parameter

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs [hunk 1]
```diff
@@ -2,17 +2,43 @@
...
-			text.text = $"---------------------------------------------------\n\nDate: {entryData.date} \n\rAuthor: {entryData.author} " +
-			 $"\n\r\n\rCommit: {entryData.commit} \n\r\n\rMessage: {entryData.message}";
+			text.text = $"{CatagoryGrabber(entryData.category)} - {entryData.description}\n\rAuthor: {entryData.author_username}";
+		}
+
+		private string CatagoryGrabber(string catagory)
+		{
+			switch (catagory)
+			{
+				case "New":
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs]: I'm pretty sure the category string is in all caps

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs [hunk 2]
```diff
@@ -2,17 +2,43 @@
...
+			public static string NEW => "";
+			public static string IMPROVE => "";
+			public static string UNKNOWN => "?";
+		}
 
 		public void SetEntry(ChangeLogEntry entryData)
 		{
-			text.text = $"---------------------------------------------------\n\nDate: {entryData.date} \n\rAuthor: {entryData.author} " +
-			 $"\n\r\n\rCommit: {entryData.commit} \n\r\n\rMessage: {entryData.message}";
+			text.text = $"{CatagoryGrabber(entryData.category)} - {entryData.description}\n\rAuthor: {entryData.author_username}";
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs]: why not link to the pr and author?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs]: was still using text at that time, not tmp

Comment (corp-0) [reply UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs]: aight, leave it as is. I will make a pr to your branch with some changes

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs]: done

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs [hunk 3]
```diff
@@ -11,23 +19,61 @@ public class ChangeLogHandler : MonoBehaviour
...
+				return;
+			}
+			GrabJsonData();
+		}
+
+		[Button("Get Data")]
+		private void GrabJsonData()
+		{
+			var client = new HttpClient();
+			client.BaseAddress = new Uri(ChangeLogLink);
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs]: Can you leave the `BaseAddress` `null` and just pass the entire URL into `GetAsync`?

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs [hunk 4]
```diff
@@ -11,23 +19,61 @@ public class ChangeLogHandler : MonoBehaviour
...
+			// Add an Accept header for JSON format.
+			client.DefaultRequestHeaders.Accept.Add(
+				new MediaTypeWithQualityHeaderValue("application/json"));
+			// Get data response
+			var response = client.GetAsync(ChangeLogParams).Result;
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs]: Won't this block the whole game until the request is finished?

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs [hunk 5]
```diff
@@ -11,23 +19,61 @@ public class ChangeLogHandler : MonoBehaviour
...
+			response.EnsureSuccessStatusCode();
+
+			//if we failed getting the data
+			if (response.IsSuccessStatusCode == false)
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs]: This if statement will never be used, as `EnsureSuccessStatusCode` throws an exception.

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs [hunk 6]
```diff
@@ -11,23 +19,61 @@ public class ChangeLogHandler : MonoBehaviour
...
+			}
+			//ReadAsAsync<>() has been deprecated long ago so use ReadAsStringAsync() instead
+			string responseBody = response.Content.ReadAsStringAsync().Result;
+			//Read the data from the json then add it to the entries list
+			StringReader dataString = new StringReader(responseBody);
+			allEntries.AddRange(ReadFromJson(dataString).results);
+			//avoid memory leaks by disposing all of this
+			dataString.Dispose();
+			response.Dispose();
+			client.Dispose();
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogHandler.cs]: `using` statements prevent leakage by always calling dispose (ex. when exceptions happen).

[Reference](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement)

File: UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs [hunk 7]
```diff
@@ -2,17 +2,43 @@
...
+		}
+
+		private string CatagoryGrabber(string catagory)
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/UI/Core/ChangeLogs/ChangeLogEntryUI.cs]:
```suggestion
		private string CategoryGrabber(string category)
```

**Review by Alainx277 — CHANGES_REQUESTED** [review]

**Comment by corp-0** [comment]
removed myself as reviewer because I wrote commits for it

**Comment by MaxIsJoe** [comment]
can still request a review from you

File: UnityProject/Assets/Scripts/UI/Systems/ServerInfoPanel/Pages/ChangelogPage.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+			ChangeDataPage(changelogData.Value.previous);
+		}
+
+		private static async Task<AllChangesResponse> FetchChanges(string url = BASE_API_URL)
+		{
+			AllChangesResponse newData = null;
+			using var client = new HttpClient();
+			client.DefaultRequestHeaders.Accept.Add(
+				new MediaTypeWithQualityHeaderValue("application/json"));
+			using var response = client.GetAsync(url).Result;
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/UI/Systems/ServerInfoPanel/Pages/ChangelogPage.cs]: This still blocks on the web request.

```suggestion
			using var response = await client.GetAsync(url);
```

File: UnityProject/Assets/Scripts/Util/SweetExtensions.cs [hunk 9]
```diff
@@ -3,30 +3,30 @@
...
+using System.Threading.Tasks;
 using Items;
 using Messages.Server;
 
 public static class SweetExtensions
 {
 	public static Pickupable PickupableOrNull(this GameObject go)
 	{
-		return go.OrNull()?.GetComponent<Pickupable>();
+		return go.OrNull().OrNull()?.GetComponent<Pickupable>();
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Util/SweetExtensions.cs]: Why does this need `OrNull` twice?

**Review by Alainx277 — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/Util/SweetExtensions.cs (hunk 9 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Util/SweetExtensions.cs]: It doesn't, my mistake

File: UnityProject/Assets/Scripts/UI/Systems/ServerInfoPanel/Pages/ChangelogPage.cs (hunk 8 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/UI/Systems/ServerInfoPanel/Pages/ChangelogPage.cs]: Aight, gonna fix it

**Review by Alainx277 — APPROVED** [review]

**Pull request merged**